clients/src/main/java/org/apache/kafka/common/serialization/ListDeserializer.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.serialization;
+
+import com.sun.xml.internal.ws.encoding.soap.DeserializationException;
+import org.apache.kafka.clients.CommonClientConfigs;
+import org.apache.kafka.common.utils.Utils;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ListDeserializer<T> implements Deserializer<List<T>> {
+
+    private Deserializer<T> inner;
+    private Class listClass;
+    private Integer primitiveSize;
+
+    private Map<Class, Integer> primitiveDeserializers = new HashMap<Class, Integer>() {{
+        put(LongDeserializer.class, 8);
+        put(IntegerDeserializer.class, 4);
+        put(ShortDeserializer.class, 2);
+        put(FloatDeserializer.class, 4);
+        put(DoubleDeserializer.class, 8);
+    }};
+
+    public ListDeserializer() {
+    }
+
+    public ListDeserializer(Class listClass, Deserializer<T> deserializer) {
+        this.listClass = listClass;
+        this.inner = deserializer;
+        this.primitiveSize = primitiveDeserializers.get(deserializer.getClass());
+    }
+
+    @Override
+    public void configure(Map<String, ?> configs, boolean isKey) {
+        if (inner == null) {
+            String listTypePropertyName = isKey ? CommonClientConfigs.DEFAULT_LIST_KEY_SERDE_TYPE_CLASS : CommonClientConfigs.DEFAULT_LIST_VALUE_SERDE_TYPE_CLASS;
+            String innerSerdePropertyName = isKey ? CommonClientConfigs.DEFAULT_LIST_KEY_SERDE_INNER_CLASS : CommonClientConfigs.DEFAULT_LIST_VALUE_SERDE_INNER_CLASS;
+            String listType = (String) configs.get(listTypePropertyName);
```

Comment:
I think this could be `String` or `Class` type. Not sure. For any case, we should test for both cases.

Reply:
You mean we can directly cast it to `Class` object?
i.e. `Class listType = (Class) configs.get(listTypePropertyName);`

Reply:
Yes. The user can use the config two ways:
```
// as string
props.put(DEFAULT_LIST_KEY_SERDE_TYPE_CLASS, "my.fully.qualified.package.MyInnerSerde");
// or as class
props.put(DEFAULT_LIST_KEY_SERDE_TYPE_CLASS, my.fully.qualified.package.MyInnerSerde.class);
// or short (it the class is imported)
props.put(DEFAULT_LIST_KEY_SERDE_TYPE_CLASS, MyInnerSerde.class);
```

Both should be supported and the code need to be able to handle both cases. Hence, we should get is as `Object` and use `instanceof` to check the type.

Reply:
Something like this I presume:
```
final Object innerSerde = configs.get(innerSerdePropertyName);
try {
    if (innerSerde instanceof String) {
        inner = Utils.newInstance((String) innerSerde, Serde.class).serializer();
    } else if (innerSerde instanceof Class) {
        inner = ((Serde<T>) Utils.newInstance((Class) innerSerde)).serializer();
    } else {
        throw new ClassNotFoundException();
    }
    inner.configure(configs, isKey);
} catch (final ClassNotFoundException e) {
    throw new ConfigException(innerSerdePropertyName, innerSerde, "Serde class " + innerSerde + " could not be found.");
}
```

Reply:
Not 100% sure -- but we need tests for this cases. The `configure()` code is untested atm